2.4/root/usr/share/container-scripts/mongodb/common.sh

```diff
@@ -119,25 +119,60 @@ function mongo_initiate() {
 
 # get the address of the current primary member
 function mongo_primary_member_addr() {
-  local current_endpoints
-  current_endpoints=$(endpoints)
-  local mongo_node
-  mongo_node="$(echo "${current_endpoints}" | grep -v "$(container_addr)" | head -1):${CONTAINER_PORT}"
-  echo -n $(mongo admin -u admin -p "${MONGODB_ADMIN_PASSWORD}" --host ${mongo_node} --quiet --eval "print(rs.isMaster().primary);")
+  endpoints | grep -v "$(container_addr)" | while read mongo_node; do
+    cmd_output="$(mongo admin -u admin -p "$MONGODB_ADMIN_PASSWORD" --host "$mongo_node:$CONTAINER_PORT" --eval 'print(rs.isMaster().primary)' --quiet || true)"
+
+    # Trying to find IP:PORT in output and filter out error message because mongo prints it to stdout
+    ip_and_port_regexp='[0-9]\+\.[0-9]\+\.[0-9]\+\.[0-9]\+:[0-9]\+'
+    if addr="$(echo "$cmd_output" | grep -x "$ip_and_port_regexp")"; then
+      echo -n "$addr"
+      exit 0
+    fi
+
+    echo >&2 "Cannot get address of primary from $mongo_node node: $cmd_output"
+    exit 1
+  done && return
+
+  echo >&2 "Cannot get address of primary node: after checking all nodes we don't have the address"
+
+  # return empty string
+  echo -n ''
 }
 
 # mongo_remove removes the current MongoDB from the cluster
 function mongo_remove() {
-  echo "=> Removing $(mongo_addr) on $(mongo_primary_member_addr) ..."
+  local primary_addr
+  primary_addr="$(mongo_primary_member_addr)"
+
+  if [ -z "$primary_addr" ]; then
+    echo >&2 "Cannot remove node from the replica set: primary node was not found"
+    return
+  fi
+
+  local mongo_addr
+  mongo_addr="$(mongo_addr)"
+
+  echo "=> Removing $mongo_addr on $primary_addr ..."
   mongo admin -u admin -p "${MONGODB_ADMIN_PASSWORD}" \
-    --host $(mongo_primary_member_addr) --eval "rs.remove('$(mongo_addr)');" &>/dev/null || true
+    --host "$primary_addr" --eval "rs.remove('$mongo_addr');" || true
 }
 
 # mongo_add advertise the current container to other mongo replicas
 function mongo_add() {
-  echo "=> Adding $(mongo_addr) to $(mongo_primary_member_addr) ..."
+  local primary_addr
+  primary_addr="$(mongo_primary_member_addr)"
+
+  if [ -z "$primary_addr" ]; then
+    echo >&2 "Cannot add node to the replica set: primary node was not found"
+    return
```

Comment:
Shouldn't we `exit 1` here instead of returning?

When `mongo_primary_member_addr` returns nothing, that means `mongo_add` should fail and we should stop immediately.

Otherwise we'd had printed an error to stderr, but still proceed to execute whatever came after `mongo_add`.

The same for `mongo_remove`.